mitiq/calibration/calibration.py

```diff
@@ -0,0 +1,204 @@
+                    circuit, expval_executor
+                )
+                mitigated[strategy.technique]["results"].append(
+                    {
+                        "circuit_type": circuit_data.type,
+                        "mitigated_value": mitigated_expval,
+                        **strategy.technique_params,
+                    }
+                )
+            expvals.append(
+                {
+                    "noisy_dist": noisy_distribution,
+                    "ideal_dist": ideal_distribution,
+                    "mitigated": mitigated,
+                }
+            )
+        return expvals
+
+    def compute_improvements(
+        self, experiment_results: list[dict[str, Any]]
+    ) -> None:
+        """Compute the improvement factors for each calibration circuit that
+        was run."""
+        for result in experiment_results:
+            ideal_dist = result["ideal_dist"]
+            ideal_expval = max(ideal_dist.values())
+            noisy_dist = result["noisy_dist"]
+            noisy_expval = max(noisy_dist.values())
+            for di in result["mitigated"].values():
+                results = di["results"]
+                mitigated_vals = list(
+                    map(lambda di: di["mitigated_value"], results)
+                )
+                method_improvement_factor = abs(
+                    noisy_expval - ideal_expval
+                ) / sqrt(
+                    len(mitigated_vals)
+                    * sum(
+                        (mitigated_val - ideal_expval) ** 2
+                        for mitigated_val in mitigated_vals
+                    )
+                )
+                di["method_improvement_factor"] = method_improvement_factor
+
+    def get_optimal_strategy(self, results: list[dict[str, Any]]) -> str:
+        """Finds the optimal error mitigation strategy using the improvement
+        factors calculated, and stored in `self.results`."""
+        best_val = 0.0
+        best_key = ""
+        for result in results:
+            ideal_dist = result["ideal_dist"]
+            ideal = max(ideal_dist.values())
+            noisy_dist = result["noisy_dist"]
+            noisy = max(noisy_dist.values())
+            unmitigated_error = abs((ideal - noisy) / ideal)
+            for method, di in result["mitigated"].items():
+                for res in di["results"]:
+                    mitigated_expval = res["mitigated_value"]
+                    diff = abs((ideal - mitigated_expval) / ideal)
+                    error_diff = abs(unmitigated_error - diff)
+                    if error_diff > best_val:
+                        best_val = error_diff
+                        best_key = method
+        return best_key
+
+    def run(self) -> str:
+        results = self.run_circuits()
+        self.compute_improvements(results)
+        self.results = results
+        strategy = self.get_optimal_strategy(results)
+        return strategy
```

Comment:
```suggestion
     return results
```
Do you think `run()` should compute the optimal strategy? Maybe it could just run the circuits and compute the improvement factors. The optimal strategy could depend on the specific real use-case circuit, so we may postpone its computation at a later step.

Reply:
Agreed. I had it as more of a place holder, but you're right that it's not needed here.